prebuild/emscripten/post.js

```diff
@@ -0,0 +1,1010 @@
+  }
+};;
+
+P2.prototype['dbl'] = P2.prototype.dbl = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  return wrapPointer(_emscripten_bind_P2_dbl_0(self), P2);
+};;
+
+  P2.prototype['__destroy__'] = P2.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  _emscripten_bind_P2___destroy___0(self);
+};
+// SecretKey
+/** @suppress {undefinedVars, duplicate} @this{Object} */function SecretKey() {
+  this.ptr = _emscripten_bind_SecretKey_SecretKey_0();
+  getCache(SecretKey)[this.ptr] = this;
+};;
+SecretKey.prototype = Object.create(WrapperObject.prototype);
+SecretKey.prototype.constructor = SecretKey;
+SecretKey.prototype.__class__ = SecretKey;
+SecretKey.__cache__ = {};
+Module['SecretKey'] = SecretKey;
+
+SecretKey.prototype['keygen'] = SecretKey.prototype.keygen = /** @suppress {undefinedVars, duplicate} @this{Object} */function(ikm, info) {
+  var self = this.ptr;
+  var ikm_len = 0;
+  ensureCache.prepare();
+  if (ikm && typeof ikm === 'object') {
+    if (ikm instanceof Uint8Array) {
+      ikm_len = ikm.length;
+      ikm = ensureUint8(ikm, ikm.length);
+    } else {
+      throw new Error("unsupported ikm type");
+    }
+  } else {
+    ikm_len = ikm.length;
+    ikm = ensureString(ikm);
+  }
+  if (info && typeof info === 'object') info = info.ptr;
+  else info = ensureString(info || "");
+  _emscripten_bind_SecretKey_keygen_3(self, ikm, ikm_len, info);
+};;
+
+SecretKey.prototype['from_bendian'] = SecretKey.prototype.from_bendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function(input) {
+  var self = this.ptr;
+  ensureCache.prepare();
+  if (typeof input == 'object') { input = ensureInt8(input); }
+  _emscripten_bind_SecretKey_from_bendian_1(self, input);
+};;
+
+SecretKey.prototype['from_lendian'] = SecretKey.prototype.from_lendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function(input) {
+  var self = this.ptr;
+  ensureCache.prepare();
+  if (typeof input == 'object') { input = ensureInt8(input); }
+  _emscripten_bind_SecretKey_from_lendian_1(self, input);
+};;
+
+SecretKey.prototype['to_bendian'] = SecretKey.prototype.to_bendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  ensureCache.prepare();
+  var out = ensureCache.alloc(new Uint8Array(32), HEAPU8);
+  // var out = _malloc(32);
+  _emscripten_bind_SecretKey_to_bendian_1(self, out);
+  return new Uint8Array(HEAPU8.subarray(out, out + 32));
+  // TODO: free `out`?
+};;
+
+SecretKey.prototype['to_lendian'] = SecretKey.prototype.to_lendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  ensureCache.prepare();
+  var out = ensureCache.alloc(new Uint8Array(32), HEAPU8);
+  _emscripten_bind_SecretKey_to_lendian_1(self, out);
+  return new Uint8Array(HEAPU8.subarray(out, out + 32));
+  // TODO: free `out`?
```

Comment:
My understanding is that Uint8Array(wasm-pointer, ...) provides direct access to wasm memory. My question is HEAP8.subarray() makes a copy of the data from wasm memory to the Javascript pool. (I apologize if the question is stupid,  I'm not that familiar with interactions between wasm and Javascript.)

Reply:
The Internet suggests that one might just as well return a pointer to the webasm stack, which you then immediately copy to the Javascript pool. I suppose this means that data sits safely on the stack till the next call from to the corresponding webasm module. This way you don't need to allocate a temporary buffer and figure out when it can be freed...

Reply:
> My question is [if] HEAP8.subarray() makes a copy of the data from wasm memory to the Javascript pool.

From what I gather `.subarray` doesn't, but in the context `new Uint8Array` does the copy. As for freeing `out`. My understanding is that `ensureCache` allocates a chunk of wasm memory and reuses it specifically for argument passing. It has no free method.